packages/atclient/src/atclient.c

```diff
@@ -270,14 +270,136 @@ exit: { return ret; }
 }
 
 int atclient_get_publickey(const atclient *atclient, const atclient_atkey *atkey, char *value, const size_t valuelen,
-                           size_t *valueolen) {
+                           size_t *valueolen, bool bypasscache) {
   int ret = 1;
 
-  // TODO: implement
+  // 1. initialize variables
+  atclient_atstr cmdbuffer;
+  atclient_atstr_init_literal(&cmdbuffer, 4096, "plookup:");
```

Comment:
I don't think we need that much space (4096 bytes are a lot). Perhaps we could also use some temporal arrays (`char command[len];`), as @XavierChanth suggested in a previous PR, we are not going to use them outside this function.

Reply:
I prefer:
- count number of bytes needed (or approximate if size difference is negligible)
- allocate the bytes
- use it

over:
- allocate really big buffer
- use it

The first uses slightly more cpu in exchange for lower memory usage, and we are more concerned with memory right now.

Reply:
This was "fixed" here I think: https://github.com/atsign-foundation/at_c/commit/23737d4cb224619c3f40678a1faeeeb43d558900

My cmdbuffer size looks like:

```c
const size_t cmdbufferlen = strlen("plookup:all:\r\n") + (bypasscachestr != NULL ? strlen(bypasscachestr) : 0) + strlen(atkeystrwithoutpublic) + 1;
cmdbuffer = malloc(sizeof(char) * cmdbufferlen);
```

is there a way I can avoid malloc'ing for these two variables (below) as well? I am allocating 4096 twice here again
```c
// 1. initialize variables
const size_t atkeystrlen = ATKEY_GENERAL_BUFFER_SIZE;
atclient_atstr atkeystr;
atclient_atstr_init(&atkeystr, atkeystrlen);

const size_t recvlen = 4096;
atclient_atbytes recv;
atclient_atbytes_init(&recv, recvlen);
```

But I am thinking that these two are unaviodable because 1. the atkey string output of the to_string operation is unknown and the output of the atclient_recv is also unknown, so we will need to allocate a large size everytime to avoid overflow

Reply:
If you want to avoid the malloc, for the first, you can just:

```
const size_t cmdbufferlen = strlen("plookup:all:\r\n") + (bypasscachestr != NULL ? strlen(bypasscachestr) : 0) + strlen(atkeystrwithoutpublic) + 1;
char cmdbuffer[cmdbufferlen]; 
```
This way you don't have to worry about doing free on exit. Also you aren't initializing the `cmdbuffer` until the moment you are going to use it, so the change shouldn't break anything. Just remove the null init at the top.

For the second two cases:
1. Here it is comfortable to use an `atclient_atstr` because its parameters perfectly fit the output of `atclient_atkey_to_string`:
https://github.com/atsign-foundation/at_c/blob/25c9de562375740fd96ec9770a07020c7dcd68bc/packages/atclient/include/atclient/atkey.h#L64-L65
But you could also do it without it:
```
const size_t atkeystrlen = ATKEY_GENERAL_BUFFER_SIZE;
char atkeystr[atkeystrlen];
size_t olen = 0;

...

// 2. build plookup: command
ret = atclient_atkey_to_string(*atkey, atkeystr, atkeystrlen, &olen);
if (ret != 0) {
  atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_atkey_to_string: %d\n", ret);
  goto exit;
}
```
Here `olen` isn't actually very important. Why? Let's see the `strnstr` [man](https://manpages.ubuntu.com/manpages/trusty/man3/strnstr.3.html):
```
The strnstr() function locates the first occurrence of the null-terminated string little in
the string big, where not more than len characters are searched.  Characters that appear
after a ‘\0’ character are not searched.  

// Hope this isnt a problem... If it is, we can always go back to strstr
Since the strnstr() function is a FreeBSD specific
API, it should only be used when portability is not a concern.
```
It doesn't keep looking after '\0'. Besides that, in the next line we are relying on the variable ending in '\0' as well (because we are doing `strlen(atkeystrwithoutpublic)`:
https://github.com/atsign-foundation/at_c/blob/23737d4cb224619c3f40678a1faeeeb43d558900/packages/atclient/src/atclient.c#L315
But it is okay to pass it to `strnstr` anyways. If it turns out we cant use `strnstr` and we have to use `strstr` then it wont ask for that 'n'.

2. Similarly:
```
const size_t recvlen = 4096;
unsigned char recv[recvlen];
size_t recvolen = 0;    // You can pass this to atclient_stringutils_starts_with
```

Reply:
@realvarx I think doing `char cmdbuffer[cmdbufferlen];` does indeed cause an error:

Error:
![image](https://github.com/atsign-foundation/at_c/assets/79019866/12805be8-042b-41bf-965a-f7a4e0f3a40a)

That is because before this allocation, it is possible that we do `goto exit`

So I do think that the null pointer at the top is necessary because then it can be freed later on

Reply:
But `char cmdbuffer[cmdbufferlen];` is static memory. You dont have to free it in the exit

Reply:
I can't use static memory here because of the `goto exit` branch

I think this illustrates the reason why I can't use static memory in this case:

For example:

```c
#include <stdio.h>

int main() {

    if(0) {
        goto exit;
    }

    size_t len = 4;
    char cmdbuffer[len];
exit: {
    return 0;
}
}
```

I get this error:
![image](https://github.com/atsign-foundation/at_c/assets/79019866/42ab6c3a-7a22-4067-b7ee-2440a2768561)

Since there's a branch (`goto exit`) **before** the allocation of cmdbuffer, we can't use static memory allocation and have to resort to dynamic memory allocation.

There are 2 solutions to fixing this error:

The solution is to either

1. Move cmdbuffer to before the goto exit;

Like:
```c
size_t len = 4;
char cmdbuffer[len];
    
if(0) {
    goto exit;
}
```

But what if size_t depends on values like strlen(atkeystr) which can only be evaluated after a goto exit statement

So solution 2 is to use malloc

2. Use malloc
3. 
```c
if(0) {
    goto exit;
}

size_t len = 4;
char *cmdbuffer = malloc(len);
```

Reply:
@realvarx Let me know what you think

Reply:
@XavierChanth Would like your thoughts on this as well